src/canister/individual_user_template/src/util/migration.rs

```diff
@@ -0,0 +1,203 @@
+use std::{borrow::BorrowMut, collections::BTreeMap, error::Error};
+
+use candid::{CandidType, Principal};
+use ic_cdk::{
+    api::{
+        call, canister_balance128,
+        management_canister::main::{canister_info, CanisterInfoRequest},
+    },
+    call, caller, id,
+};
+use serde::{Deserialize, Serialize};
+use shared_utils::{
+    canister_specific::individual_user_template::types::{
+        migration::{MigrationErrors, MigrationInfo},
+        post::{Post, PostDetailsFromFrontend},
+    },
+    common::{
+        types::{known_principal::KnownPrincipalType, utility_token::token_event::TokenEvent},
+        utils::system_time::{get_current_system_time, get_current_system_time_from_ic},
+    },
+};
+
+use crate::{
+    api::post::add_post_v2::{self, add_post_to_memory},
+    data_model::CanisterData,
+    CANISTER_DATA,
+};
+
+#[derive(PartialEq)]
+pub enum SubnetType {
+    HotorNot,
+    Yral,
+}
+
+pub trait Migration {
+    async fn transfer_tokens_and_posts(
+        &self,
+        caller: Principal,
+        to_individual_user: IndividualUser,
+    ) -> Result<(), MigrationErrors>;
+
+    fn recieve_tokens_and_posts(
+        &self,
+        from_individual_user: IndividualUser,
+        token_amount: u64,
+        posts: BTreeMap<u64, Post>,
+    ) -> Result<(), MigrationErrors>;
+}
+
+pub struct IndividualUser {
+    pub canister_id: Principal,
+    pub profile_principal: Principal,
+    pub subnet_type: SubnetType,
+    pub migration_status: Option<MigrationInfo>,
+}
+
+impl IndividualUser {
+    pub async fn from_canister_data() -> Result<Self, MigrationErrors> {
+        let (profile_principal, migration_info) = CANISTER_DATA.with_borrow(|canister_data| {
+            let Some(profile_principal) = canister_data.profile.principal_id else {
+                return Err(MigrationErrors::UserNotRegistered);
+            };
+            Ok((profile_principal, canister_data.migration_info))
+        })?;
+
+        IndividualUser::new(id(), profile_principal, Some(migration_info)).await
+    }
+
+    pub async fn new(
+        canister_id: Principal,
+        profile_principal: Principal,
+        migration_info: Option<MigrationInfo>,
+    ) -> Result<IndividualUser, MigrationErrors> {
+        let (canister_info,) = canister_info(CanisterInfoRequest {
+            canister_id,
+            num_requested_changes: None,
+        })
+        .await
+        .map_err(|_e| MigrationErrors::CanisterInfoFailed)?;
+
+        let hot_or_not_subnet_orchestrator_canister_id =
+            CANISTER_DATA.with_borrow(|canister_data| {
+                canister_data
+                    .known_principal_ids
+                    .get(&KnownPrincipalType::CanisterIdHotOrNotSubnetOrchestrator)
```

Comment:
Are we sure KnownPrincipalType::CanisterIdHotOrNotSubnetOrchestrator is present in all individual canisters ?

Reply:
No we will use platform-orchestrator to issue global id that will reflect in all canisters

Reply:
which method is this ?